matchmaking.py

```diff
@@ -0,0 +1,78 @@
+import random
+import time
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+class Matchmaking:
+    def __init__(self, li, config, username):
+        self.li = li
+        self.variants = list(filter(lambda variant: variant != "fromPosition", config["challenge"]["variants"]))
+        self.matchmaking_cfg = config.get("matchmaking") or {}
+        self.username = username
+        self.last_challenge_created = time.time()
+        self.challenge_expire_time = 25  # The challenge expires 20 seconds after creating it.
+        self.challenge_id = None
+
+    def should_create_challenge(self):
+        matchmaking_enabled = self.matchmaking_cfg.get("allow_matchmaking")
+        time_has_passed = self.last_challenge_created + ((self.matchmaking_cfg.get("challenge_interval") or 30) * 60) < time.time()
+        challenge_expired = self.last_challenge_created + self.challenge_expire_time < time.time() and self.challenge_id
+        # Wait 10 seconds before creating a new challenge to avoid hitting the api rate limits.
+        ten_seconds_passed = self.last_challenge_created + 10 < time.time()
+        if challenge_expired:
+            self.li.cancel(self.challenge_id)

```

Comment:
After this line probably something like:
```py
logger.info(f"Challenge id {self.challenge_id} cancelled.")
```
could be added?

Reply:
I added it with debug instead of info, because this line is only useful in correspondence challenges, because challenges are automatically canceled after 20 seconds except for correspondence challenges. After 25 seconds we send cancel_challenge to cancel correspondence challenges.